File: gauntlet/agents/safety_gate.py
"""
Deterministic Action Safety Gate.
This is a pure code check (NOT an LLM call) verifying entity provenance and grounding.
"""

from typing import Tuple, List
from gauntlet.schemas.contracts import LeadOutput, BuilderOutput, SafetyGateResult

def evaluate_safety_gate(raw_dump: str, lead_output: LeadOutput, builder_output: BuilderOutput) -> SafetyGateResult:
    """
    Deterministically audits all entities referenced in Builder artifacts:
    1. Verifies that every Lead entity has a verbatim source_span in the raw dump.
    2. Verifies that every entity referenced by Builder belongs to verified Lead entities.
    """
    normalized_dump = raw_dump.lower()
    
    # 1. Verify Lead extracted entity source spans against raw dump
    verified_known_values = set()
    for e in lead_output.entities:
        span = (e.source_span or e.value or "").strip().lower()
        if span and span in normalized_dump:
            verified_known_values.add(e.value.strip().lower())
        elif e.value.strip().lower() in normalized_dump:
            verified_known_values.add(e.value.strip().lower())
            
    # 2. Collect all referenced entities across builder artifacts
    verified_refs = []
    unverified_refs = []
    
    for artifact in builder_output.artifacts:
        for ref in artifact.referenced_entities:
            ref_clean = ref.strip()
            ref_lower = ref_clean.lower()
            if not ref_lower:
                continue
                
            matched = False
            for known in verified_known_values:
                if ref_lower in known or known in ref_lower:
                    matched = True
                    break
                    
            if matched or ref_lower in normalized_dump:
                if ref_clean not in verified_refs:
                    verified_refs.append(ref_clean)
            else:
                if ref_clean not in unverified_refs:
                    unverified_refs.append(ref_clean)
                    
    total = len(verified_refs) + len(unverified_refs)
    score = 100 if total == 0 else int((len(verified_refs) / total) * 100)
    passed = len(unverified_refs) == 0
    
    summary = (
        f"Grounding verified: All {len(verified_refs)} action entities proven from source notes."
        if passed
        else f"Safety Gate blocked {len(unverified_refs)} unverified entities: {', '.join(unverified_refs)}"
    )
    
    return SafetyGateResult(
        passed=passed,
        grounding_score=score,
        verified_entities=verified_refs,
        unverified_entities=unverified_refs,
        audit_summary=summary,
    )

File: src/lib/gauntlet/safety-gate.ts
/**
 * Stage 4: Deterministic Action Safety Gate
 *
 * This is a pure, zero-LLM deterministic code check that guarantees grounding:
 * 1. Verifies that every extracted entity has a verbatim source_span matching the original raw dump.
 * 2. Verifies that all referenced entities across the builder's artifacts belong to the verified source entities.
 *
 * Zero hallucination, deterministic, zero-latency execution.
 */

import type { ExtractedEntity, SafetyGateReport } from "./types";
import type { BuilderArtifact } from "./agents/builder";

export function evaluateSafetyGate(
  rawDump: string,
  extractedEntities: ExtractedEntity[],
  artifacts: BuilderArtifact[],
): SafetyGateReport {
  const normalizedDump = rawDump.toLowerCase();

  // 1. Verify that extracted entities actually exist in the original raw dump
  const verifiedSourceEntities: ExtractedEntity[] = [];
  const invalidSourceEntities: ExtractedEntity[] = [];

  for (const entity of extractedEntities) {
    const span = (entity.source_span || entity.value || "").trim().toLowerCase();
    if (span.length > 0 && normalizedDump.includes(span)) {
      verifiedSourceEntities.push({ ...entity, verified: true });
    } else {
      // Fallback: check if the value itself is present
      const val = (entity.value || "").trim().toLowerCase();
      if (val.length > 0 && normalizedDump.includes(val)) {
        verifiedSourceEntities.push({ ...entity, verified: true });
      } else {
        invalidSourceEntities.push({ ...entity, verified: false });
      }
    }
  }

  const knownValues = new Set(
    verifiedSourceEntities.map((e) => e.value.trim().toLowerCase()),
  );

  // 2. Check referenced entities from artifacts
  const verifiedRefs = new Set<string>();
  const unverifiedRefs = new Set<string>();

  for (const artifact of artifacts) {
    const refs = artifact.referenced_entities || [];
    for (const ref of refs) {
      const normalizedRef = ref.trim().toLowerCase();
      if (!normalizedRef) continue;

      let matched = false;
      for (const known of knownValues) {
        if (normalizedRef.includes(known) || known.includes(normalizedRef)) {
          matched = true;
          break;
        }
      }

      if (matched || normalizedDump.includes(normalizedRef)) {
        verifiedRefs.add(ref);
      } else {
        unverifiedRefs.add(ref);
      }
    }
  }

  const unverifiedList = Array.from(unverifiedRefs);
  const verifiedList = Array.from(verifiedRefs);

  // Calculate grounding score
  const totalRefs = verifiedList.length + unverifiedList.length;
  const groundingScore =
    totalRefs === 0
      ? 100
      : Math.round((verifiedList.length / totalRefs) * 100);

  const passed = unverifiedList.length === 0;

  const audit_summary = passed
    ? `Grounding verified: All ${verifiedSourceEntities.length} entities and ${verifiedList.length} action references traced back to raw source notes.`
    : `Safety Gate Flagged ${unverifiedList.length} ungrounded entities not found in original notes: ${unverifiedList.join(", ")}`;

  return {
    passed,
    score: groundingScore,
    verified_entities: verifiedList.length > 0 ? verifiedList : verifiedSourceEntities.map((e) => e.value),
    unverified_entities: unverifiedList,
    audit_summary,
  };
}
